### DISCUSSION TITLE
Combining conditions

### DISCUSSION THREAD

**Discussion by franckvallee** [evrete/evrete#17] [Ideas]
I would expect that using JAVA boolean operators in @Where would be possible.
ex:
>@Where( "$x.age > 18 && $x.nationality.equals(\'French\" )"

Unfortunately this is not possible.

When combining multiple conditions like:
>@Where( {"$x.age > 18", "$x.nationality.equals(\'French\"} )"

Is that considered as a AND between conditions?
How to implement an OR without using the asMethods declaration?

**Comment by dmfay** [comment]
Multiple conditions in a @Where are AND-ed together. I think you'd have to use asMethods for an OR if you want to avoid duplicating rule bodies.

**Comment by andbi** [comment]
As for the OR operator, each literal condition is compiled as a stand-alone static Java method. Thus one can use the standard `||` operator inside conditions.

**Comment by code-uri** [comment]
How do create a rule with nested conditions like below. at a root level I can combine multiple AND's using where() method. 
But, how can I implement the below scenario? 

```
condition 1 -> ( condition 3 OR condition 4)
AND
condition 2 -> condition 5 && condition 6
```

If the conditions are simple string expressions,  I can join the conditions using || ,  && operators. But, what if the conditions are functions ?

**Reply by andbi** [reply]
Functions can also be joined with boolean operators and grouped with parentheses.

Alternatively, if such a condition refers properties of a single type, sometimes it's easier to declare a new boolean field on this type and use it as `$a.myNewField`

**Reply by code-uri** [reply]
You mean, I can include any arbitrary static class function as a string in my expression something like this?
` "$x.flag == true && ( DateUtils.before($x.timestamp,  '2023-03-17') || DateUtils.after($x.timestamp,  '2024-03-17') "`

**Reply by andbi** [reply]
Using static function in literal conditions is possible too. 

The engine compiles conditions into Java classes, and it is possible for a developer to provide a base class for all of them. Inside that base class you can statically import or define any functions and use them in your conditions.

**Comment by code-uri** [comment]
Thank you @andbi, adding my utils classes to imports in  configuration  worked.

**Reply by andbi** [reply]
Yes, using imports is another way to tackle this task. Somehow it slipped my mind, thanks for a reminder. 

However, if you find your conditions hard to read or debug, you can always create a parent utility class for all your conditions, see the `evrete.impl.condition-base-class` property in the docs https://evrete.org/docs/configuration/
